### ISSUE TITLE
Rule `require-default-props` fails on custom validator

### ISSUE THREAD

**Issue by mminer** [yannickcr/eslint-plugin-react#1020]
For prop types that use a custom validator, there's no straightforward way (that I'm aware of) to tell React that it's required. For these cases, the `require-default-props` rule correctly reports that no default is specified, but this isn't particularly useful — I *can't* apply `isRequired`, even though that's the intent.

Ideally the `require-default-props` rule would ignore custom prop validators, making it unnecessary to sprinkle `eslint-disable` comments wherever they appear.

```javascript
MyComponent.propTypes = {
  customProp: (props, propName) => {
    ...
  }, // <- Unable to specify `isRequired`, so require-default-props reports it
};
```

**Comment by ljharb** [comment]
Usually I define all my custom propTypes in a separate file, and attach an `isRequired` to them, mirroring the facebook ones (https://www.npmjs.com/package/airbnb-prop-types is an example).

However, for inline ones, it's true that there's not really a way to know if they're required or not.

It'd be good to be able to opt into this checking tho - i'm just not sure how.

**Comment by lwassermann** [comment]
Although it is rather painful, there is a way to work around this, which sacrifices both readability and understandability:
```javascript
MyComponent.propTypes = {
  customProp: { isRequired(props, propName) {
      ...
    }
  }.isRequired,
};
```
…

**Comment by ljharb** [comment]
Technically it's a bug that this plugin cares about `.isRequired` on non-react propTypes, since it's impossible to know non-react propTypes' semantics there.

I don't think there's any way with static analysis that we can tell if something's required unless it's an actual React propType, and the current `.isRequired` hack seems fine to keep.
